### PR TITLE
Faster `torch.Tensor.shape[0]` to get length for `torch.Tensor`

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Standardizes tensor dimension access across Ultralytics to use .shape[...] and improves empty-tensor checks, enhancing robustness and consistency across YOLO and SAM workflows. 🧹✅

### 📊 Key Changes
- Replaced len(tensor) and .size(0) with .shape[0] throughout the codebase (SAM, YOLO detect/segment/pose/classify/OBB, metrics, nn, tasks).
- Adjusted conditions to use tensor.shape[0] for empty checks (e.g., masks, predictions) instead of len(...), ensuring correct behavior with tensors.
- Updated tensor expansions and indexing to use .shape (e.g., expand, assertions, view).
- Minor fixes:
  - augment.get_visuals initializes tensors using cls_unique.shape[0].
  - Corrected a docstring type in segment predict (pred is torch.Tensor).
  - Small view fix in nn/modules/block.py using self.gamma.shape[0].
- Consistency tweaks in SAM decoders and predictors for batch handling and stability selection.

### 🎯 Purpose & Impact
- Improves reliability with PyTorch tensors, especially for scripted/export contexts and non-list structures. 🛡️
- Reduces potential edge-case bugs when inputs are empty or have non-standard dimensions. 🧠
- Increases code clarity and consistency across modules, making maintenance easier. 🧩
- No API changes; performance impact is negligible; behavior should be the same or more robust in edge cases. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help speed things up — an Ultralytics engineer will review and assist you shortly 🙂

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR:
- Since this replaces `len(tensor)` with `tensor.size(0)` across the codebase, please ensure export/tracing paths (TorchScript/ONNX) and edge cases (empty tensors) are covered in tests.
- If any behavioral changes are expected in validators, predictors, or metrics, consider adding small unit tests to demonstrate correctness across detection/segmentation/pose/OBB tasks.
- Keep diffs focused and consistent with existing coding style to simplify review 🙌

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by Laughing-q** [comment]
actually in my test `shape` with index is even faster than torch.size
```python
def test_numel():
    import torch
    import timeit

    x = torch.randn(16, 3, 640, 640).cuda()
    x_np = x.cpu().numpy()
    def torch_size():
        # return x.numel()
        return x.size(0)

    def torch_python_len():
        return len(x)
        # return x.shape[1]

    def torch_shape():
        # return x.numel()
        return x.shape[0]

    def numpy_shape():
        return x_np.shape[0]

    def numpy_python_len():
        return len(x_np)

    print("torch size:", timeit.timeit(torch_size, number=100000))
    print("torch python len:", timeit.timeit(torch_python_len, number=100000))
    print("torch shape:", timeit.timeit(torch_shape, number=100000))
    print("numpy shape:", timeit.timeit(numpy_shape, number=100000))
    print("numpy python len:", timeit.timeit(numpy_python_len, number=100000))

test_numel()
```
```text
torch size: 0.011774022999816225
torch python len: 0.032942472000286216
torch shape: 0.009677386999101145
numpy shape: 0.004644567001378164
numpy python len: 0.0023563319991808385
```

**Review by glenn-jocher — APPROVED** [review]

**Comment by glenn-jocher** [comment]
@Laughing-q wow this is really nice optimization! This could have lots of effect as the ops are used a lot.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged! Huge thanks to @Laughing-q for leading and to @glenn-jocher for the contributions in [PR #22021](https://github.com/ultralytics/ultralytics/pull/22021).

“Simplicity is the ultimate sophistication.” — Leonardo da Vinci  
Your shift to consistent `.shape[...]` access and robust empty-tensor checks brings that spirit to our codebase—cleaner, safer, and more maintainable across YOLO and SAM workflows. This kind of thoughtful refinement elevates reliability today and speeds future innovation. We appreciate your craftsmanship and attention to detail! 🙌
